### PR TITLE
WPS based Data Export

### PR DESCRIPTION
**Pull request by vlt1**
## Description
WPS based data export implementation. Contains WPS download client.

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [ ] Bugfix
 - [x] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [ ] Other... Please describe:

## Issue
#6144 
#6156 
#6157 

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

### CODE REVIEW

**Comment by vlt1** [comment]
I will now proceed with adding docs and tests first for #6156 then for #6157

**Comment by offtherailz** [comment]
@vlt1 can you solve the conflicts ? (we migrated syntax from require to import) they should be very easy to solve.

File: web/client/translations/data.it-IT.json [hunk 1]
```diff
@@ -1610,16 +1610,43 @@
...
             "format": "Formato file",
             "srs": "Sistema di riferimento",
             "export": "Esporta",
             "downloadonlycurrentpage": "Scarica solo la pagina corrente",
+            "cropDataSet": "Ritaglia il set di dati nella finestra corrente",
+            "downloadFilteredDataSet": "Scarica set di dati filtrati",
+            "advancedOptions": "Opzioni avanzate",
+            "compression": "Compressione",
+            "tileWidth": "Larghezza piastrella",
+            "tileHeight": "Altezza piastrella",
```
Comment (offtherailz) [inline web/client/translations/data.it-IT.json:1624]:
```suggestion
            "tileWidth": "Larghezza tile",
            "tileHeight": "Altezza tile",
```

File: web/client/product/plugins.js [hunk 2]
```diff
@@ -118,8 +120,7 @@ export default {
         UserManagerPlugin: require('../plugins/manager/UserManager').default,
         UserExtensionsPlugin: require('../plugins/UserExtensions').default,
         UserSessionPlugin: require('../plugins/UserSession').default,
-        VersionPlugin: require('../plugins/Version').default,
-        WFSDownloadPlugin: require('../plugins/WFSDownload').default,
+        VersionPlugin: require('../plugins/Version'),
```
Comment (offtherailz) [inline web/client/product/plugins.js:123]:
```suggestion
        VersionPlugin: require('../plugins/Version').default,
```

**Review by offtherailz — CHANGES_REQUESTED** [review]
- [ ] see my inline comments
- [ ] when WPS is missing, so using ordinary WFS service the download doesn't work anymore (the `typeName` is sent as `undefined` to the server)
```xml
<wfs:GetFeature service="WFS" version="2.0" xmlns:wfs="http://www.opengis.net/wfs/2.0" xmlns:fes="http://www.opengis.net/fes/2.0" xmlns:gml="http://www.opengis.net/gml/3.2" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.opengis.net/wfs/2.0 http://schemas.opengis.net/wfs/2.0/wfs.xsd http://www.opengis.net/gml/3.2 http://schemas.opengis.net/gml/3.2.1/gml.xsd"><wfs:Query typeNames="undefined" ></wfs:Query></wfs:GetFeature>
```
- [ ] Actually if I have WPS I have this list of formats, that is a big limitation over the normal WFS download list. 
image 

the normal list is this. 

image.  

Moreover, proposing these 4 formats for every layer maybe confusing, because only a sub-set of them is available, depending on the feature type. @giohappy is this the desired behaviour ? 


Minor things:
- [ ] Instead of display "no results found" if the download list is empty, we should show something more clear, like. "Export list is empty"
- [ ] Maybe the two plugins should be merged in one single plugin (the ExportResultsList can not live without download and vice-versa) to simplify the management. 
- [ ] Maybe the download status button should not be always present, but only when I have some downloads in the list.
image

**Comment by vlt1** [comment]
> Actually if I have WPS I have this list of formats, that is a big limitation over the normal WFS download list.

The formats in WPS list are the ones that actually work with WPS.

> Moreover, proposing these 4 formats for every layer maybe confusing, because only a sub-set of them is available, depending on the feature type.

How do I determine if a particular format is available for a specific layer?

**Comment by offtherailz** [comment]
Thanks @vlt1 for clarification. @tdipisa is this acceptable or should we somehow allow to switch to "WFS only mode" or disable WPS option with a flag? 
One more question for @vlt1 : where the list is stored, to recover it in a second time after browser reload?

**Comment by tdipisa** [comment]
@vlt1 few more comments below.

**1.** 
>Actually if I have WPS I have this list of formats, that is a big limitation over the normal WFS download list.

At the moment there is no direct way in GeoServer to retrieve the list of mime types supported by the WPS Download process (**this could be an interesting improvement to do in GeoServer in my opinion**). The supported mime types list can be also related to eventual additional extensions in GeoServer (like DXF). For this reason, now we need a default configuration list of mime types defined for the download plugin in MapStore. This configuration must be different for vector and raster layers. Below the list we can use for the default configuration in MapStore:

Vector layers (**a proper label need to be associated to all of them in the MS UI**):

```
- text/xml; subtype=wfs-collection/1.0
- text/xml; subtype=wfs-collection/1.1
- application/wfs-collection-1.0
- application/wfs-collection-1.1
- application/zip
- text/csv
- application/json
```

Raster layers (**a proper label need to be associated to all of them in the MS UI**):

```
- image/tiff 
- application/arcgrid
- image/png
- image/jpeg
```

**2.**
When activating the MS tool for the Download process, we have to check if the layer is a raster layer or a vector layer and then provide only the corresponding list of formats (one of the above lists depending on the layer).

At the moment there is no clean way to understand if a layer is a vector or a raster. Available options are the following ones:

**Option 1**
a. While activating the MS tool, perform a WCS GetCapabilities request to see if the layer is reported there: if yes, the layer is a raster layer
b. If the layer is not reported by GeoServer in the WCS GetCapabilities output, perform a second WFS GetCapabilities request to see if the layer is reported there: if yes, the layer is a vector layer

_**The limitation here are:**_
- Possible performance issues 
- The WCS and WFS can be disabled in GeoServer (in this case the MS tool should inform the user that the Download is not possible) 

**Option 2**
a. While adding a layer from the MS Catalog tool (WMS source), check if the keyword WPS is present and put this information in the layer state so that MS is able to reconize if the layer is a raster or not and so provide the corresponding mime types list accordingly.

_**The limitation here are:**_
- This is a weak check because usually the keyword WPS is automatically set for raster layers by GeoServer, but the keyword can be removed/changed from the GeoServer layer configuration panel

Providing a better way to check if a layer is vector or raster could be a really interesting improvement for GeoServer.
**For the moment, in my opinion, we should opt for Option 2** (as I already told you @vlt1).

**3.**
The download tool must not be available in some cases (e.g. for local vector layers like annotations). Also for this we have to do a preventive check to enable the tool only for WFS/WMS layers (this is an information we have in MS).

File: web/client/plugins/LayerDownload.jsx [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+import { createPlugin } from '../utils/PluginsUtils';
+
+/**
+ * Provides advanced export functionalities using WFS and WPS.
+ * @memberof plugins
+ * @name LayerDownload
+ * @class
+ * @prop {object[]} formats An array of name-label objects for the allowed formats available.
+ * @prop {object[]} srsList An array of name-label objects for the allowed srs available. Use name:'native' to omit srsName param in wfs filter
+ * @prop {string} defaultSrs Deafult selected srs
```
Comment (offtherailz) [inline web/client/plugins/LayerDownload.jsx:57]:
```suggestion
 * @prop {string} defaultSrs Default selected srs
```

**Review by offtherailz — APPROVED** [review]
I think is very good. :+1:

**Pull request merged**